### PR TITLE
pypolicyd-spf 2.0.2 -> spf-engine 2.9.3

### PR DESCRIPTION
**Pull request by yorickvP**
Fixes CVE-2019-20790 (#90974)

###### Description of changes

pypolicyd-spf was renamed to spf-engine. [Fedora made a similar change.](https://src.fedoraproject.org/rpms/pypolicyd-spf/c/5e1fce533559a7421828c8ca7d8dcdbf402023fd?branch=rawhide).

[Changelog](https://git.launchpad.net/spf-engine/tree/CHANGES?h=2.9.3&id=405688f17f376d134cfa1fa97f5e853de87210bb)

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/servers/mail/spf-engine/default.nix [hunk 1]
```diff
@@ -1,12 +1,12 @@
 { lib, buildPythonApplication, fetchurl, pyspf }:
 
 buildPythonApplication rec {
-  pname = "pypolicyd-spf";
-  majorVersion = "2.0";
-  version = "${majorVersion}.2";
+  pname = "spf-engine";
+  majorVersion = builtins.head (builtins.match "([0-9]+\.[0-9]+)\..*" version);
```
Comment (risicle) [inline pkgs/servers/mail/spf-engine/default.nix:5]: We have version handling functions in `lib/versions.nix`

**Comment by risicle** [comment]
`nixpkgs-review` happy, macos 10.15, nixos x86_64 & aarch64.

File: pkgs/servers/mail/spf-engine/default.nix [hunk 2]
```diff
@@ -1,12 +1,12 @@
...
+  version = "2.9.3";
 
   src = fetchurl {
-    url = "https://launchpad.net/pypolicyd-spf/${majorVersion}/${version}/+download/${pname}-${version}.tar.gz";
+    url = "https://launchpad.net/${pname}/${majorVersion}/${version}/+download/${pname}-${version}.tar.gz";
     sha256 = "1nm8y1jjgx6mxrbcxrbdnmkf8vglwp0wiw6jipzh641wb24gi76z";
```
Comment (risicle) [inline pkgs/servers/mail/spf-engine/default.nix:10]: Hold up - the hash hasn't changed - we're still building the old package.

Comment (yorickvP) [reply pkgs/servers/mail/spf-engine/default.nix]: Nice catch!

**Comment by yorickvP** [comment]
Updated the hash, had to add the `pymilter` package, but now it's actually 2.9.3 :)

**Comment by yorickvP** [comment]
nixpkgs-review still happy

File: pkgs/development/python-modules/pymilter/default.nix [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+  buildInputs = [ libmilter ];
+  propagatedBuildInputs = [ bsddb3 pydns ];
+
+  preBuild = ''
+    sed -i 's/import thread/import _thread as thread/' Milter/greylist.py
+  '';
+
+  # requires /etc/resolv.conf
+  doCheck = false;
```
Comment (risicle) [inline pkgs/development/python-modules/pymilter/default.nix:22]: We should at least have a `pythonImportsCheck` if we're not going to have the tests enabled.

FWIW see the `libredirect` package and its various uses around nixpkgs for how to provide "fake" files at arbitrary filesystem locations for test suites.

**Comment by risicle** [comment]
@ofborg build spf-engine

Fails for me on macos 10.15

**Comment by yorickvP** [comment]
Not sure what to do about that error. I've asked for backup and disabled the darwin pymilter build for now.

**Comment by risicle** [comment]
The macos failure is actually an indication that `libmilter` is slightly broken there. To fix it, it needs this alteration:

```
--- a/pkgs/development/libraries/libmilter/default.nix
+++ b/pkgs/development/libraries/libmilter/default.nix
@@ -1,4 +1,4 @@
-{ lib, stdenv, fetchurl, m4 }:
+{ lib, stdenv, fetchurl, m4, fixDarwinDylibNames }:
 
 stdenv.mkDerivation rec {
   pname = "libmilter";
@@ -34,7 +34,11 @@ stdenv.mkDerivation rec {
 
   patches = [ ./install.patch ./sharedlib.patch ./glibc-2.30.patch ./darwin.patch ];
 
-  nativeBuildInputs = [ m4 ];
+  nativeBuildInputs = [ m4 fixDarwinDylibNames ];
+
+  postInstall = lib.optionalString stdenv.isDarwin ''
+    fixDarwinDylibNames $out/lib/libmilter.dylib.1
+  '';
 
   meta = with lib; {
     description = "Sendmail Milter mail filtering API library";
```

and then the `libredirect` setup just needs to be altered to instead set `DYLD_PRELOAD` to `libredirect.dylib` in the darwin case.

Alternatively you can add `libredirect` to the `buildInputs` and let its setup hook do all the preload-setting for you.

**Comment by yorickvP** [comment]
Adding fixDarwinDylibNames seems to work. However, adding `libredirect` to `checkInputs` does nothing. Adding `DYLD_INSERT_LIBRARIES` makes all files fail with 'permission denied'. Doing nothing works since macos *has* `/etc/resolv.conf` (:

File: nixos/doc/manual/release-notes/rl-2211.section.md [hunk 4]
```diff
@@ -235,6 +235,8 @@ Available as [services.patroni](options.html#opt-services.patroni.enable).
 - `dockerTools.buildImage` deprecates the misunderstood `contents` parameter, in favor of `copyToRoot`.
   Use `copyToRoot = buildEnv { ... };` or similar if you intend to add packages to `/bin`.
 
+- The `pypolicyd-spf` package was updated from 2.0.2 to 2.9.3 and renamed to `spf-engine`.
+
```
Comment (SuperSandro2000) [inline nixos/doc/manual/release-notes/rl-2211.section.md:239]:
```suggestion
```
Not note worthy.

File: pkgs/servers/mail/spf-engine/default.nix [hunk 5]
```diff
@@ -0,0 +1,26 @@
+{ lib, buildPythonApplication, fetchurl, pyspf, dnspython, authres, pymilter }:
+
+buildPythonApplication rec {
+  pname = "spf-engine";
+  majorVersion = lib.versions.majorMinor version;
```
Comment (SuperSandro2000) [inline pkgs/servers/mail/spf-engine/default.nix:5]:
```suggestion
```

File: pkgs/servers/mail/spf-engine/default.nix [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+  version = "2.9.3";
+
+  src = fetchurl {
+    url = "https://launchpad.net/${pname}/${majorVersion}/${version}/+download/${pname}-${version}.tar.gz";
```
Comment (SuperSandro2000) [inline pkgs/servers/mail/spf-engine/default.nix:9]:
```suggestion
    url = "https://launchpad.net/spf-engine/${lib.versions.majorMinor version}/${version}/+download/spf-engine-${version}.tar.gz";
```

File: pkgs/servers/mail/spf-engine/default.nix [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+    sha256 = "sha256-w0Nb+L/Os3KPApENoylxCVaCD4FvgmvpfVvwCkt2IDE=";
+  };
+
+  propagatedBuildInputs = [ pyspf dnspython authres pymilter ];
+
+  preBuild = ''
```
Comment (SuperSandro2000) [inline pkgs/servers/mail/spf-engine/default.nix:15]:
```suggestion
  postPatch = ''
```

File: pkgs/servers/mail/spf-engine/default.nix [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+    substituteInPlace setup.py --replace "'/etc'" "'$out/etc'"
+  '';
+
+  meta = with lib; {
+    homepage = "https://launchpad.net/${pname}/";
+    description = "Postfix policy engine for Sender Policy Framework (SPF) checking";
+    maintainers = with maintainers; [ abbradar ];
+    license = licenses.asl20;
+    platforms = platforms.all;
```
Comment (SuperSandro2000) [inline pkgs/servers/mail/spf-engine/default.nix:24]:
```suggestion
```
already defaults to linux and darwin

File: pkgs/servers/mail/spf-engine/default.nix [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+  };
+
+  propagatedBuildInputs = [ pyspf dnspython authres pymilter ];
+
+  preBuild = ''
+    substituteInPlace setup.py --replace "'/etc'" "'$out/etc'"
+  '';
+
+  meta = with lib; {
+    homepage = "https://launchpad.net/${pname}/";
```
Comment (SuperSandro2000) [inline pkgs/servers/mail/spf-engine/default.nix:20]:
```suggestion
    homepage = "https://launchpad.net/spf-engine/";
```

File: pkgs/servers/mail/spf-engine/default.nix [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+  majorVersion = lib.versions.majorMinor version;
+  version = "2.9.3";
+
+  src = fetchurl {
+    url = "https://launchpad.net/${pname}/${majorVersion}/${version}/+download/${pname}-${version}.tar.gz";
+    sha256 = "sha256-w0Nb+L/Os3KPApENoylxCVaCD4FvgmvpfVvwCkt2IDE=";
+  };
+
+  propagatedBuildInputs = [ pyspf dnspython authres pymilter ];
+
```
Comment (SuperSandro2000) [inline pkgs/servers/mail/spf-engine/default.nix:14]: Can we run tests here?

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]
Please add pythonImportChecks everywhere.

File: pkgs/servers/mail/spf-engine/default.nix (hunk 10 above)
Comment (yorickvP) [reply pkgs/servers/mail/spf-engine/default.nix:13]: I don't see a test runner

**Pull request merged**